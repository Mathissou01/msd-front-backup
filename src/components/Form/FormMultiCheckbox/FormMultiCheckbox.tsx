import React, { useEffect } from "react";
import { useFormContext } from "react-hook-form";
import { ErrorMessage } from "@hookform/error-message";
import classNames from "classnames";
import CommonFormErrorText from "../../Common/CommonFormErrorText/CommonFormErrorText";
import FormLabel from "../FormLabel/FormLabel";
import "./form-multicheckbox.scss";

interface IOption {
  label: string;
  value: string;
  defaultChecked?: boolean;
}

interface IFormMultiCheckboxProps {
  name: string;
  label?: string;
  secondaryDisplayName?: string;
  isRequired?: boolean;
  isDisabled?: boolean;
  options: Array<IOption>;
  defaultValue?: string;
  displayMode?: "vertical" | "horizontal";
  onChange?: (data: unknown) => void;
}

export default function FormMultiCheckbox({
  name,
  label,
  secondaryDisplayName,
  isRequired = false,
  isDisabled = false,
  options,
  defaultValue,
  displayMode = "horizontal",
  onChange,
}: IFormMultiCheckboxProps) {
  /* Static Data */
  const errorMessages = {
    required: "Ce champ est obligatoire",
  };

  /* Local Data */
  const {
    register,
    watch,
    setValue,
    formState: { isSubmitting, errors },
  } = useFormContext();

  const watchChecked: Array<string> = watch(name, []);

  useEffect(() => {
    if (
      Array.isArray(watchChecked) &&
      watchChecked.length === 0 &&
      defaultValue
    ) {
      setValue(name, defaultValue);
    }
  }, [defaultValue, name, setValue, watchChecked]);

  useEffect(() => {
    if (watchChecked && onChange) {
      onChange(watchChecked);
    }
  }, [watchChecked, onChange]);

  return (
    <div className="c-FormMultiCheckbox">
      <fieldset>
        {label && (
          <FormLabel
            label={label}
            isRequired={isRequired}
            secondaryLabel={secondaryDisplayName}
            tagType="legend"
          />
        )}
        <div
          className={classNames("c-FormMultiCheckbox__Options", {
            "c-FormMultiCheckbox__Options_horizontal":
              displayMode === "horizontal",
            "c-FormMultiCheckbox__Options_vertical": displayMode === "vertical",
          })}
        >
          {options.map((option, index) => (
            <div key={index} className="c-FormMultiCheckbox__Option">
              <input
                className={`c-FormMultiCheckbox__Input ${
                  Array.isArray(watchChecked) &&
                  watchChecked.find((item) => item === option.value)
                    ? "c-FormMultiCheckbox__Input_checked"
                    : ""
                }`}
                type="checkbox"
                {...register(name, {
                  required: {
                    value: isRequired,
                    message: errorMessages.required,
                  },
                })}
                id={`${name}${index}`}
                value={option.value}
                defaultChecked={option.defaultChecked}
                disabled={isSubmitting || isDisabled}
                data-testid={`form-multi-checkbox_${index}`}
              />
              <FormLabel forId={`${name}${index}`} label={option.label} />
            </div>
          ))}
        </div>
      </fieldset>
      <ErrorMessage
        errors={errors}
        name={name}
        render={({ message }: { message: string }) => (
          <CommonFormErrorText message={message} errorId={`${name}_error`} />
        )}
      />
    </div>
  );
}
