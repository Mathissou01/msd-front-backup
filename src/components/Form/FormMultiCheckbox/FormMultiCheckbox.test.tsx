import React, { ReactNode } from "react";
import { FormProvider, useForm } from "react-hook-form";
import { fireEvent, render, screen } from "@testing-library/react";
import FormMultiCheckbox from "./FormMultiCheckbox";

const mock = {
  name: "form-checkbox",
  label: "form checkbox",
  options: [
    { label: "test1", value: "value1" },
    { label: "test2", value: "value2" },
    { label: "test3", value: "value3" },
    { label: "test4", value: "value4" },
  ],
};

const Wrapper = (props: { children: ReactNode }) => {
  const formMethods = useForm();
  return <FormProvider {...formMethods}>{props.children},</FormProvider>;
};

describe("FormMultiCheckbox", () => {
  it("renders", () => {
    const { container } = render(
      <Wrapper>
        <FormMultiCheckbox
          name={mock.name}
          label={mock.label}
          options={mock.options}
          isRequired
        />
      </Wrapper>,
    );

    const checkbox = screen.getByTestId("form-multi-checkbox_0");
    expect(checkbox).toBeInTheDocument();
    expect(checkbox).not.toHaveClass("c-FormMultiCheckbox__Input_checked");

    expect(container).toMatchSnapshot();
  });

  it("changes style when checked", async () => {
    render(
      <Wrapper>
        <FormMultiCheckbox
          name={mock.name}
          label={mock.label}
          options={mock.options}
          isRequired
        />
      </Wrapper>,
    );

    const checkbox = screen.getByTestId("form-multi-checkbox_0");
    expect(checkbox).not.toHaveClass("c-FormMultiCheckbox__Input_checked");
    fireEvent.click(checkbox);
    expect(checkbox).toHaveClass("c-FormMultiCheckbox__Input_checked");
  });

  it("can be disabled", () => {
    render(
      <Wrapper>
        <FormMultiCheckbox
          name={mock.name}
          label={mock.label}
          options={mock.options}
          isRequired
          isDisabled={true}
        />
      </Wrapper>,
    );

    const checkbox = screen.getByTestId("form-multi-checkbox_0");
    expect(checkbox).toHaveAttribute("disabled");
    fireEvent.click(checkbox);
    expect(checkbox).not.toHaveClass("c-FormMultiCheckbox__Input_checked");
  });
});
