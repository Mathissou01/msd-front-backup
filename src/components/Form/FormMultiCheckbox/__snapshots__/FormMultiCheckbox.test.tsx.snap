// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormMultiCheckbox renders 1`] = `
<div>
  <div
    class="c-FormMultiCheckbox"
  >
    <fieldset>
      <div
        class="c-FormLabel"
      >
        <legend
          class="c-FormLabel__LabelWrapper"
        >
          <span
            class="c-FormLabel__Label"
          >
            form checkbox *
          </span>
        </legend>
      </div>
      <div
        class="c-FormMultiCheckbox__Options c-FormMultiCheckbox__Options_horizontal"
      >
        <div
          class="c-FormMultiCheckbox__Option"
        >
          <input
            class="c-FormMultiCheckbox__Input "
            data-testid="form-multi-checkbox_0"
            id="form-checkbox0"
            name="form-checkbox"
            type="checkbox"
            value="value1"
          />
          <div
            class="c-FormLabel"
          >
            <label
              class="c-FormLabel__LabelWrapper"
              for="form-checkbox0"
            >
              <span
                class="c-FormLabel__Label"
              >
                test1
              </span>
            </label>
          </div>
        </div>
        <div
          class="c-FormMultiCheckbox__Option"
        >
          <input
            class="c-FormMultiCheckbox__Input "
            data-testid="form-multi-checkbox_1"
            id="form-checkbox1"
            name="form-checkbox"
            type="checkbox"
            value="value2"
          />
          <div
            class="c-FormLabel"
          >
            <label
              class="c-FormLabel__LabelWrapper"
              for="form-checkbox1"
            >
              <span
                class="c-FormLabel__Label"
              >
                test2
              </span>
            </label>
          </div>
        </div>
        <div
          class="c-FormMultiCheckbox__Option"
        >
          <input
            class="c-FormMultiCheckbox__Input "
            data-testid="form-multi-checkbox_2"
            id="form-checkbox2"
            name="form-checkbox"
            type="checkbox"
            value="value3"
          />
          <div
            class="c-FormLabel"
          >
            <label
              class="c-FormLabel__LabelWrapper"
              for="form-checkbox2"
            >
              <span
                class="c-FormLabel__Label"
              >
                test3
              </span>
            </label>
          </div>
        </div>
        <div
          class="c-FormMultiCheckbox__Option"
        >
          <input
            class="c-FormMultiCheckbox__Input "
            data-testid="form-multi-checkbox_3"
            id="form-checkbox3"
            name="form-checkbox"
            type="checkbox"
            value="value4"
          />
          <div
            class="c-FormLabel"
          >
            <label
              class="c-FormLabel__LabelWrapper"
              for="form-checkbox3"
            >
              <span
                class="c-FormLabel__Label"
              >
                test4
              </span>
            </label>
          </div>
        </div>
      </div>
    </fieldset>
  </div>
  ,
</div>
`;
